tools/themis/src/rules.rs

```diff
@@ -402,3 +402,52 @@ pub fn recursively_publishable(workspace: &Workspace) -> anyhow::Result<()> {
     }
     Ok(())
 }
+
+/// Ensure that top-level `Cargo.toml` does not contain superfluouos dependencies.
+pub fn no_superfluous_deps(workspace: &Workspace) -> anyhow::Result<()> {
+    let mut workspace_deps = BTreeSet::<String>::new();
+    let mut read_deps = |manifest: &toml::value::Value, table_name: &'static str| {
+        let Some(deps) = manifest.get(table_name) else { return };
+        let Some(deps) = deps.as_table() else { return };
+        for (name, dep) in deps {
+            let Some(workspace_field) = dep.get("workspace") else { continue };
+            let Some(true) = workspace_field.as_bool() else { continue };
+            workspace_deps.insert(name.clone());
+        }
+    };
+    for pkg in workspace.members.iter() {
+        read_deps(&pkg.manifest.raw, "dependencies");
+        read_deps(&pkg.manifest.raw, "dev-dependencies");
+        read_deps(&pkg.manifest.raw, "build-dependencies");
+        let Some(target) = pkg.manifest.raw.get("target") else { continue };
+        let Some(target) = target.as_table() else { continue };
+        for (_, target_manifest) in target {
+            read_deps(&target_manifest, "dependencies");
+            read_deps(&target_manifest, "dev-dependencies");
+            read_deps(&target_manifest, "build-dependencies");
+        }
+    }
+    let Some(root_deps) = workspace.manifest.read(&["workspace", "dependencies"]) else {
+        panic!("Cargo.toml has no [workspace.dependencies]?");
+    };
+    let Some(root_deps) = root_deps.as_table() else {
+        panic!("workspace.dependencies is not a table?");
+    };
+    let root_deps = BTreeSet::from_iter(root_deps.keys().into_iter().cloned());
+    let diff = root_deps.difference(&workspace_deps).collect::<Vec<_>>();
```

Comment:
Can you explain the high level of how this function works? What exactly does it compare?

Reply:
It gathers the names of all packages as defined in various `[*dependencies]` tables in workspace members into one set, then names of all packages in `[workspace.dependencies]` in the root `Cargo.toml` and if the latter has any packages that aren’t being used by the member packages, this `diff` will contain those names to cause a failure.

Examples of packages that this check has complained about are `conqueue` and `loupe` that were removed from `Cargo.toml`.

(In practice `[workspace.dependencies]` entries do not hurt anything as they are entirely ignored and not even checked for correctness, but for the same of hygiene it seems like a good idea to keep this table clean.)